internals/cli/masker/masker.go

```diff
@@ -0,0 +1,139 @@
+package masker
+
+import (
+	"io"
+	"time"
+)
+
+// Masker handles the creation and synchronization of streams that have all their writes scanned for secrets and
+// have them redacted if any matches are found. Masking of secrets is a best effort attempt. Output on all streams is
+// buffered to increase the chance of finding secrets if they are spread across multiple writes, but it cannot be
+// guaranteed that these secrets are masked. The duration bytes spend in the buffer is constant.
+//
+// Usage:
+// 1. Create a new Masker using New()
+// 2. Add one more streams using AddStream()
+// 3. Run the Start() method in a separate goroutine
+// 4. After everything has been written to the io.Writers, flush all buffers using Stop()
+type Masker struct {
+	bufferDelay time.Duration
+	sequences   [][]byte
+	frames      chan frame
+	stopChan    chan struct{}
+	err         error
+}
+
+// Options for configuring masking behavior.
+type Options struct {
+	// DisableBuffer completely disables the buffering of the masker. This increases output responsiveness
+	// but also increases the chance of a secret not being masked.
+	DisableBuffer bool
+
+	// BufferDelay is the constant duration for which input to a stream is buffered. A higher value increases
+	// the chance of secrets being detected for masking. Especially when writes have a variable delay between them,
+	// for example in the case data arrives over an unstable network connection.
+	// Defaults to 50ms if not set.
+	BufferDelay time.Duration
+
+	// FrameBufferLength is the number of frames that can be in the buffer simultaneously.
+	// If the frame buffer is full, writing to a stream blocks until there is space.
```

Comment:
Hmm. Interesting approach. I think there is definitely something to say for this. But it also introduces a problem: if you do this, there is a direct path from `Write()` to `flush()`, from which the time of the matching logic could theoretically be deduced.

Reply:
Now that the time of matching is constant, can we change this?

Reply:
I am not sure... because it can create a little weird flow/dependency in the code. 

My guess would be to implement this in the following way. Replace the channel read in `Start()` with this:
```golang
select {
	case <-m.forceFlush:
	case <-trigger.timer.C:
}
```
And replace `registerFrame()` with the following:
```
	frame := frame{
		length: l,
		stream: s,
		timer:  time.NewTimer(offset + m.bufferDelay),
	}
	select {
	case m.frames <- frame:
	default:
		m.forceFlush <- struct{}{}
		m.frames <- frame
	}
```

---

In most cases, this would lead to the timer being skipped when the `frames` channel is full. However, if `Start()` is not waiting for a timer but blocked in `trigger.stream.flush()` (for example because the stream's destination is blocked), this means a message is added to `m.forceFlush` that is not picked up directly. 

In the next iteration of `Start()` a message is popped from `m.frames`, freeing up a place in the channel and after that, the timer is also directly skipped. Skipping the timer is not necessary anymore, causing the wrong timer to be skipped. If you have >1024 streams in a Masker I think this could even lead to a deadlock (not that this is a realistic scenario, but it signals that the implementation is not correct).

---

Another solution would be to make sending a message on `m.forceFlush` non-blocking:
```golang
select {
	case m.skipFrameTimer <- struct{}{}:
	default:
}
m.frames <- frame
```

This goes wrong in another scenario: if two frames are registered at (roughly) the same time and `Start()` is waiting for a timer to expire. In this case, a timer is only canceled once, though for both frames the timer should be canceled.

--- 

My suggestion would be not to implement this right now, as I think the practical problems of not implementing it are relatively small. At the same time I have the feeling that trying to implement this, leads to some incorrect code that can cause bugs in the future.

My explanation might not be 100% clear. Mainly wrote it down as a reference to myself. . If anything needs further explanation, let's discuss it in a video call.

Reply:
Jep good call. We can always add more sophistication later. It's already pretty polished I feel and the code is elegant enough to allow for these types of changes if and when we need them.